### PR TITLE
Add validation and test for decorators

### PR DESCRIPTION
**Pull request by timotheeguerin**
fix https://github.com/Azure/cadl-azure/issues/827

- Update the decorators to use a shared target type validation.
- Add tests to make sure diagnostics are reported for invalid target and params.

### CODE REVIEW

File: packages/compiler/core/decorator-utils.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        expected: expected,
+      },
+      target,
+    });
+    return false;
+  }
+  return true;
+}
+
+function prettyValue(program: Program, value: any) {
```
Comment (bterlson) [inline packages/compiler/core/decorator-utils.ts:85]: I think we should have a compiler API for this eventually, something that pretty-prints the model and its properties and any additional metadata about it (decorators, projected state, other things).

Comment (bterlson) [reply packages/compiler/core/decorator-utils.ts:85]: I might just slam something like that in, sounds fun.

**Review by bterlson — APPROVED** [review]
Overall this seems great. A thought: would it be valuable to have a factory function for the validator that takes the decorator name and expected types? Not sure if in practice that saves any code so I leave it up to you.

**Comment by timotheeguerin** [comment]
One thing I was thinking could be having the name of the decorator in the `DecoratorContext` and the validate function take a `DecoratorContext` instead of the program so you don't have to keep passing the name.

File: packages/samples/grpc-kiosk-example/types.cadl [hunk 2]
```diff
@@ -26,7 +26,6 @@ model LatLng {
   longitude: float64;
 }
 
-@body
```
Comment (markcowl) [inline packages/samples/grpc-kiosk-example/types.cadl:29]: why this change?

**Review by markcowl — CHANGES_REQUESTED** [review]

File: packages/samples/grpc-kiosk-example/types.cadl (hunk 2 above)
Comment (timotheeguerin) [reply packages/samples/grpc-kiosk-example/types.cadl:29]: the validation added on the body decorator only allows ModelProperty as the target, Maybe this is an incorrect assumption but I am not sure what you'd expect by putting @body on a model.

**Review by markcowl — APPROVED** [review]
